Source/Layout/ASLayout.mm

```diff
@@ -216,52 +200,70 @@ - (void)setRetainSublayoutLayoutElements:(BOOL)retainSublayoutLayoutElements
     // to ensure sublayout elements are retained until the layouts are applied.
     self.retainSublayoutLayoutElements = YES;
     return self;
   }
   
   struct Context {
-    ASLayout *layout;
+    unowned ASLayout *layout;
     CGPoint absolutePosition;
   };
   
   // Queue used to keep track of sublayouts while traversing this layout in a DFS fashion.
   std::deque<Context> queue;
-  for (ASLayout *sublayout in self.sublayouts) {
+  for (ASLayout *sublayout in _sublayouts) {
     queue.push_back({sublayout, sublayout.position});
   }
   
-  NSMutableArray *flattenedSublayouts = [[NSMutableArray alloc] init];
+  auto flattenedSublayouts = [[NSMutableArray<ASLayout *> alloc] init];
   
   while (!queue.empty()) {
     const Context context = std::move(queue.front());
     queue.pop_front();
     
-    ASLayout *layout = context.layout;
-    const NSArray<ASLayout *> *sublayouts = layout.sublayouts;
-    const NSUInteger sublayoutsCount = sublayouts.count;
+    unowned ASLayout *layout = context.layout;
+    const NSUInteger sublayoutsCount = layout->_sublayouts.count;
     const CGPoint absolutePosition = context.absolutePosition;
     
     if (ASLayoutIsDisplayNodeType(layout)) {
       if (sublayoutsCount > 0 || CGPointEqualToPoint(ASCeilPointValues(absolutePosition), layout.position) == NO) {
         // Only create a new layout if the existing one can't be reused, which means it has either some sublayouts or an invalid absolute position.
-        layout = [ASLayout layoutWithLayoutElement:layout.layoutElement
-                                              size:layout.size
-                                          position:absolutePosition
-                                        sublayouts:@[]];
+        auto newLayout = [ASLayout layoutWithLayoutElement:layout->_layoutElement
+                                                      size:layout.size
+                                                  position:absolutePosition
+                                                sublayouts:@[]];
+        [flattenedSublayouts addObject:newLayout];
+      } else {
+        [flattenedSublayouts addObject:layout];
       }
-      [flattenedSublayouts addObject:layout];
-    } else if (sublayoutsCount > 0){
-      std::vector<Context> sublayoutContexts;
-      for (ASLayout *sublayout in sublayouts) {
-        sublayoutContexts.push_back({sublayout, absolutePosition + sublayout.position});
+    } else if (sublayoutsCount > 0) {
+      // Fast-reverse-enumerate the sublayouts array by copying it into a C-array and push_front'ing each into the queue.
+      unowned ASLayout *rawSublayouts[sublayoutsCount];
+      [layout->_sublayouts getObjects:rawSublayouts range:NSMakeRange(0, sublayoutsCount)];
+      for (NSInteger i = sublayoutsCount - 1; i >= 0; i--) {
+        queue.push_front({rawSublayouts[i], absolutePosition + rawSublayouts[i].position});
       }
-      queue.insert(queue.cbegin(), sublayoutContexts.begin(), sublayoutContexts.end());
```

Comment:
The bottom here (the search expansion) is far and away the biggest win. We no longer create another `vector` (heap allocation) and we remove 2 retain/release pairs per sublayout. The first is inserting into the vector then destroying the vector, the second is inserting into the queue and then removing from the queue.